invisible_cities/reco/trigger_functions.py

```diff
@@ -0,0 +1,198 @@
+        height     = peak[1:  ].max()        # TODO Full peak for N100
+        peak       = peak[ :-2][peak[:-2]>0] # Only counts above 0 are used for Q, last bin not considered TODO ONLY -1 for N100
+        charge     = peak.sum()
+        if len(peak)==0: continue
+        start_time = p[0]
+
+        # Trigger validity
+        trigger_w   = in_range(width , *wrange, left_closed=True , right_closed=False)
+        trigger_q   = in_range(charge, *qrange, left_closed=False, right_closed=False)
+        trigger_h   = height < hrange[1]
+        trigger_all = all([trigger_w, trigger_q, trigger_h])
+
+        # Baseline at the start of the pulse
+        baseline_t0 = baseline[start_time]
+
+        trigger     = TriggerInfo(event_id   , pmt_id
+                                 ,start_time , charge        , width    , height
+                                 ,trigger_q  , trigger_w     , trigger_h,   True, trigger_all  # By default multiflag is set as true
+                                 ,baseline_t0, wvf_max_height
+                                 ,-1         , -1            , -1) # Filled later in the coincidence module
+        triggers  .append(trigger)
+
+    # Extra-peaks protection
+    if multipk is not None:
+        # Evaluate if later peaks (must be over discard_width) are candidates for the multipeak protection
+        for i, t0 in zip(range(len(triggers[:-1])), triggers[:-1]):
+            new_val = t0.valid_peak
+            for t1 in triggers[i+1:]:
+                multi_candidate = ((t1.width        >=                              multi_w  ) &
+                                   (t1.q            >                               multi_q  ) &
+                                   (t1.trigger_time <= t0.trigger_time + t0.width + multi_t+2)) # Start of trigger + peak width + time window + FPGA delay (2)
+                if not multi_candidate:
+                    continue
+                # In case the later peak ends after the multipeak protection window ends
+                if (t1.trigger_time + t1.width) > (t0.trigger_time + t0.width + multi_t + 2): # Event partially contained in the multipeak window
+                    peak_slice = slice(t1.trigger_time - 1, t0.trigger_time + t0.width + multi_t + 2 + 1)
+                    peak       = signal[peak_slice]
+                    width      = len(peak) - 1 # TODO Remove -1 in N100
+                    charge     = peak[peak>0].sum()
+
+                    new_val    = (width >= multi_w) & (charge > multi_q)
+                else:
+                    new_val = False
+                if not new_val: break
+            triggers[i] = triggers[i]._replace(valid_peak=new_val, valid_all=all([triggers[i].valid_all, new_val]))
+
+    return triggers
+
+
+def retrieve_trigger_information(channel_config : Dict[int, int]
+                                ,trigger_config : Dict
+                                ) -> Callable:
```

Comment:
This looks like a city component, but maybe it just looks like one. I leave to you the decision of moving it to components.